[Cocoa][MSE] Fix potential memory management issues

youssefsoli opened the pull request:
#### 2319c56b4e437af7742f9e0f16d82f5ab9cfe80e
<pre>
[Cocoa][MSE] Fix potential memory management issues
<a href="https://bugs.webkit.org/show_bug.cgi?id=242478">https://bugs.webkit.org/show_bug.cgi?id=242478</a>

Reviewed by Eric Carlson.

* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.mm:
(WebCore::MediaPlayerPrivateMediaSourceAVFObjC::updateLastPixelBuffer):
The pixel buffer should be moved instead of copied.
(WebCore::MediaPlayerPrivateMediaSourceAVFObjC::updateDisplayLayerAndDecompressionSession):
An early return here vs. if/else is more idomatic.

* Source/WebCore/platform/graphics/avfoundation/objc/SourceBufferPrivateAVFObjC.mm:
(WebCore::SourceBufferPrivateAVFObjC::trackDidChangeSelected):
(WebCore::SourceBufferPrivateAVFObjC::notifyClientWhenReadyForMoreSamples):
A lifetime check on the caller should be done before using `this`.

Canonical link: <a href="https://commits.webkit.org/252291@main">https://commits.webkit.org/252291@main</a>
</pre>

eric-carlson commented on Source/WebCore/platform/graphics/avfoundation/objc/SourceBufferPrivateAVFObjC.mm:
```diff
@@ -863,8 +863,9 @@ static void bufferWasConsumedCallback(CMNotificationCenterRef, const void* liste
         if (m_decompressionSession) {
-            m_decompressionSession->requestMediaDataWhenReady([this, trackID] {
-                didBecomeReadyForMoreSamples(trackID);
+            m_decompressionSession->requestMediaDataWhenReady([weakThis = WeakPtr { *this }, this, trackID] {
```
I don't think it is helpful to pass `this` to avoid a single use of `weakPtr`

eric-carlson commented on Source/WebCore/platform/graphics/avfoundation/objc/SourceBufferPrivateAVFObjC.mm:
```diff
@@ -1458,8 +1459,9 @@ static void bufferWasConsumedCallback(CMNotificationCenterRef, const void* liste
         if (m_decompressionSession) {
-            m_decompressionSession->requestMediaDataWhenReady([this, trackID] {
-                didBecomeReadyForMoreSamples(trackID);
+            m_decompressionSession->requestMediaDataWhenReady([weakThis = WeakPtr { *this }, this, trackID] {
```
Ditto

eric-carlson reviewed (approved):
r=me

youssefsoli committed:
[Cocoa][MSE] Fix potential memory management issues
https://bugs.webkit.org/show_bug.cgi?id=242478

Reviewed by Eric Carlson.

* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateMediaSourceAVFObjC.mm:
(WebCore::MediaPlayerPrivateMediaSourceAVFObjC::updateLastPixelBuffer):
The pixel buffer should be moved instead of copied.
(WebCore::MediaPlayerPrivateMediaSourceAVFObjC::updateDisplayLayerAndDecompressionSession):
An early return here vs. if/else is more idomatic.

* Source/WebCore/platform/graphics/avfoundation/objc/SourceBufferPrivateAVFObjC.mm:
(WebCore::SourceBufferPrivateAVFObjC::trackDidChangeSelected):
(WebCore::SourceBufferPrivateAVFObjC::notifyClientWhenReadyForMoreSamples):
A lifetime check on the caller should be done before using `this`.

Canonical link: https://commits.webkit.org/252291@main
